File: proposed-projects/round-trip-experiment/reverse/20260909T012246Z_fbe56704/20260910T051241Z_fa3c6dd0/ported_implementation/src/gbnf.ts
import { Graph } from './grammar-graph/graph.ts';
import { ParseState } from './grammar-graph/parse-state.ts';
import type { UnresolvedRule, ValidInput } from './grammar-graph/types.ts';
import { buildRuleStack } from './grammar-parser/build-rule-stack.ts';
import { RulesBuilder } from './rules-builder/rules-builder.ts';
import { GrammarParseError } from './utils/errors/grammar-parse-error.ts';

export const GBNF = (input: string, initialString: ValidInput = ''): ParseState => {
  const grammar = typeof input === 'string' ? input : String(input);
  const { rules, symbolIds } = new RulesBuilder(grammar);
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  if (!symbolIds.has('root')) {
    throw new GrammarParseError(
      grammar,
      0,
      `Grammar does not contain a root symbol. Available symbols are: [${symbolIds
        .keys()
        .map((key) => JSON.stringify(key))
        .join(', ')}]`
    );
  }
  const rootId = symbolIds.get('root');

  // a plain loop, rather than `map`, so that holes left by undefined rules are visited
  const stackedRules: UnresolvedRule[][][] = [];
  for (let ruleId = 0; ruleId < rules.length; ruleId++) {
    stackedRules.push(buildRuleStack(rules[ruleId]));
  }
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

export default GBNF;

File: packages/gbnf-experiment/data/20260909T194429Z_a7bb3ecf/ported_implementation/gbnf.py
from __future__ import annotations

import json
from typing import List, Optional

from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_graph.types import UnresolvedRule, ValidInput
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder import RulesBuilder
from .utils.errors.grammar_parse_error import GrammarParseError


def GBNF(input: str, initial_string: ValidInput = '') -> ParseState:
    grammar = input if isinstance(input, str) else str(input)
    builder = RulesBuilder(grammar)
    rules, symbol_ids = builder.rules, builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, 'No rules were found')
    if symbol_ids.get('root') is None:
        raise GrammarParseError(
            grammar,
            0,
            'Grammar does not contain a root symbol. '
            f'Available symbols are: {json.dumps({})}',
        )
    root_id = symbol_ids.get('root')

    stacked_rules: List[Optional[List[List[UnresolvedRule]]]] = [
        build_rule_stack(rule) if rule is not None else None for rule in rules
    ]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))
